py/desiutil/bitmask.py

```diff
@@ -0,0 +1,165 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+"""
+================
+desiutil.bitmask
+================
+
+Mask bits for the spectro pipeline.
+
+Individual packages will define their own mask bits and use this as a utility
+access wrapper.  Typical users will not need to construct BitMask objects on
+their own.
+
+Stephen Bailey, Lawrence Berkeley National Lab
+Fall 2015
+
+Example::
+
+    #  Creating a BitMask
+    from desiutil.bitmask import BitMask
+
+    import yaml
+    _bitdefs = yaml.load('''
+    ccdmask:
+        - [BAD,       0, "Pre-determined bad pixel (any reason)"]
+        - [HOT,       1, "Hot pixel"]
+        - [DEAD,      2, "Dead pixel"]
+        - [SATURATED, 3, "Saturated pixel from object"]
+        - [COSMIC,    4, "Cosmic ray"]
+    ''')
+    ccdmask = BitMask('ccdmask', _bitdefs)
+
+    #  Accessing the mask
+    ccdmask.COSMIC | ccdmask.SATURATED  #  2**4 + 2**3
+    ccdmask.mask('COSMIC')     #  2**4, same as ccdmask.COSMIC
+    ccdmask.mask(4)            #  2**4, same as ccdmask.COSMIC
+    ccdmask.COSMIC             #  2**4, same as ccdmask.mask('COSMIC')
+    ccdmask.bitnum('COSMIC')   #  4
+    ccdmask.bitname(4)         #  'COSMIC'
+    ccdmask.names()            #  ['BAD', 'HOT', 'DEAD', 'SATURATED', 'COSMIC']
+    ccdmask.names(3)           #  ['BAD', 'HOT']
+    ccdmask.comment(0)         #  "Pre-determined bad pixel (any reason)"
+    ccdmask.comment('COSMIC')  #  "Cosmic ray"
+"""
+
+
+class _MaskBit(int):
+    """
+    A single mask bit.  Subclasses int to act like an int, but allows the
+    ability to extend with blat.name, blat.comment, blat.mask, blat.bitnum.
+    """
+    def __new__(cls, name, bitnum, comment, extra=dict()):
+        self = super(_MaskBit, cls).__new__(cls, 2**bitnum)
+        self.name = name
+        self.bitnum = bitnum
+        self.mask = 2**bitnum
+        self.comment = comment
+        for key, value in extra.items():
+            assert key not in (
+                'bitlength', 'conjugate', 'real', 'imag',
+                'numerator', 'denominator'), \
+                "key '{}' already in use by int objects".format(key)
+            self.__dict__[key] = value
+        return self
+
+    def __str__(self):
+        return '{:16s} bit {} mask 0x{:X} - {}'.format(
+            self.name, self.bitnum, self.mask, self.comment)
+
+
+#  Class to provide mask bit utility functions
+class BitMask(object):
+    """BitMask object.
+    """
+    def __init__(self, name, bitdefs):
+        """
+        Args:
+            name : name of this mask, must be key in bitdefs
+            bitdefs : dictionary of different mask bit definitions;
+                each value is a list of [bitname, bitnum, comment]
+
+        Typical users are not expected to create BitMask objects directly.
```

Comment:
Then how *are* users expected to create BitMask objects?  In the example in the module's docstring, you show a BitMask object being created directly.